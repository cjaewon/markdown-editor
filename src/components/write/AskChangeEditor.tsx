import * as React from 'react';
import PopupOKCancel from '../common/PopupOKCancel';
import { WriteMode } from '../../modules/write';

export interface AskChangeEditorProps {
  visible: boolean;
  onCancel: () => void;
  onConfirm: () => void;
  convertTo: WriteMode;
}

const AskChangeEditor: React.FC<AskChangeEditorProps> = ({
  visible,
  onCancel,
  onConfirm,
  convertTo,
}) => {
  if (convertTo === WriteMode.MARKDOWN) {
    return (
      <PopupOKCancel
        visible={visible}
        title="마크다운 에디터로 전환"
        onCancel={onCancel}
        onConfirm={onConfirm}
      >
        에디터 모드를 전환하시겠습니까?
        {/* <br />
        확인을 누르시면 지금까지 입력하신 모든 내용이 마크다운 포맷으료
        변환됩니다.
        <br />
        <br />
        <div style={{ fontSize: '0.875rem' }}>
          <b>주의:</b> 변환 후 코드블록의 언어 타입을 수동으로 입력해주셔야
          합니다.
        </div> */}
      </PopupOKCancel>
    );
  }
  return (
    <PopupOKCancel
      visible={visible}
      title="쉬운 에디터로 전환"
      onCancel={onCancel}
      onConfirm={onConfirm}
    >
      에디터 모드를 전환하시겠습니까?
    </PopupOKCancel>
  );
};

export default AskChangeEditor;
