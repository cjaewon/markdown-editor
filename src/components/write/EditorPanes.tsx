import * as React from 'react';
import styled, { css } from 'styled-components';

import transitions from '../../lib/styles/transitions';
import media from '../../lib/styles/media';
const EditorPanesBlock = styled.div`
  width: 100%;
  height: 100%;
  display: flex;
  animation: ${transitions.fadeIn} 0.5s forwards;
`;
const EditorPane = styled.div<{ shadow?: boolean }>`
  min-width: 0;
  flex: 1;
  display: flex;
  flex-direction: column;
  position: relative;
  ${props =>
    props.shadow &&
    css`
      z-index: 1;
      box-shadow: 0px 0px 8px rgba(0, 0, 0, 0.015);
    `}
`;

const RightPane = styled(EditorPane)`
  ${media.medium} {
    display: none;
  }
`;

export interface EditorPanesProps {
  theme?: 'LIGHT' | 'DARK';
  left?: React.ReactNode;
  right?: React.ReactNode;
}

const EditorPanes: React.FC<EditorPanesProps> = ({
  theme = 'LIGHT',
  left,
  right,
}) => {
  return (
    <EditorPanesBlock>
      <EditorPane
        shadow
        data-testid="left"
        style={{
          backgroundColor: theme === 'DARK' ? '#263238' : 'white',
        }}
      >
        {left}
      </EditorPane>
      <RightPane
        data-testid="right"
        style={{
          backgroundColor: theme === 'DARK' ? 'white' : '#fbfdfc',
        }}
      >
        {right}
      </RightPane>
    </EditorPanesBlock>
  );
};

export default EditorPanes;
