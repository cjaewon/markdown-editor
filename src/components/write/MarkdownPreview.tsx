import * as React from 'react';
import styled from 'styled-components';
import MarkdownRender from '../common/MarkdownRender';
import palette from '../../lib/styles/palette';

const MarkdownPreviewBlock = styled.div`
  word-break: break-word;
  padding: 3rem;
  flex: 1;
  overflow-y: auto;
  color: ${palette.gray9};
`;

const Title = styled.h1`
  font-size: 2.5em;
  margin-bottom: 4rem;
  font-weight: 800;
`;

export interface MarkdownPreviewProps {
  markdown: string;
  title: string;
  className?: string;
}

const MarkdownPreview: React.FC<MarkdownPreviewProps> = ({
  markdown,
  title,
  className,
}) => {
  return (
    <MarkdownPreviewBlock id="preview">
      <Title>{title}</Title>
      <MarkdownRender markdown={markdown} editing />
    </MarkdownPreviewBlock>
  );
};

export default MarkdownPreview;
