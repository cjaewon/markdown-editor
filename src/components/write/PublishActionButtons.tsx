import * as React from 'react';
import styled from 'styled-components';
import Button from '../common/Button';
import media from '../../lib/styles/media';

const PublishActionButtonsBlock = styled.div`
  display: flex;
  justify-content: flex-end;
  ${media.custom(767)} {
    margin-top: 2rem;
  }
`;

export interface PublishActionButtonsProps {
  onCancel: () => void;
  onPublish: () => void;
  edit: boolean;
}

const PublishActionButtons: React.FC<PublishActionButtonsProps> = ({
  onCancel,
  onPublish,
  edit,
}) => {
  return (
    <PublishActionButtonsBlock>
      <Button
        size="large"
        data-testid="cancelPublish"
        color="gray"
        onClick={onCancel}
      >
        취소
      </Button>
      <Button size="large" data-testid="publish" onClick={onPublish}>
        {edit ? '수정하기' : '출간하기'}
      </Button>
    </PublishActionButtonsBlock>
  );
};

export default PublishActionButtons;
