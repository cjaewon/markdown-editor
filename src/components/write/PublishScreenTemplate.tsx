import * as React from 'react';
import styled, { css } from 'styled-components';
import palette from '../../lib/styles/palette';
import zIndexes from '../../lib/styles/zIndexes';
import transitions from '../../lib/styles/transitions';
import HideScroll from '../common/HideScroll';
import media from '../../lib/styles/media';

const PublishScreenTemplateBlock = styled.div<{ visible: boolean }>`
  display: flex;
  justify-content: center;
  align-items: center;
  position: fixed;
  @media (max-width: 1024px) and (orientation: landscape) {
    align-items: flex-start;
    padding-top: 2rem;
    padding-bottom: 2rem;
    overflow: auto;
  }

  ${media.custom(767)} {
    align-items: flex-start;
    padding-top: 2rem;
    padding-bottom: 2rem;
    overflow: auto;
    padding-left: 1rem;
    padding-right: 1rem;
  }

  left: 0;
  top: 0;
  width: 100%;
  height: 100%;
  background: ${palette.gray0};
  z-index: ${zIndexes.PublishScreen};
  animation: ${transitions.slideUp} 0.25s forwards ease-in;
  ${props =>
    props.visible
      ? css`
          animation: ${transitions.slideUp} 0.25s forwards ease-in;
        `
      : css`
          animation: ${transitions.slideDown} 0.125s forwards ease-out;
        `}
`;

const Wrapper = styled.div`
  width: 768px;
  display: flex;
  ${media.medium} {
    width: 704px;
  }
  ${media.custom(767)} {
    flex-direction: column;
  }
`;
const Pane = styled.div`
  flex: 1;
  min-width: 0;
`;
const RightPane = styled(Pane)`
  display: flex;
  flex-direction: column;
  justify-content: space-between;
`;
const Separator = styled.div`
  width: 1px;
  min-height: 425px;
  background: ${palette.gray2};
  margin-left: 2rem;
  margin-right: 2rem;
  ${media.custom(767)} {
    display: none;
  }
`;

export interface PublishScreenTemplateProps {
  visible: boolean;
  left: React.ReactNode;
  right: React.ReactNode;
}

const { useState, useEffect } = React;
const PublishScreenTemplate: React.FC<PublishScreenTemplateProps> = ({
  visible,
  left,
  right,
}) => {
  const [animate, setAnimate] = useState(false);

  useEffect(() => {
    let timeoutId: null | number = null;
    if (visible) {
      setAnimate(true);
    } else if (!visible && animate) {
      timeoutId = setTimeout(() => {
        setAnimate(false);
      }, 125);
    }
    return () => {
      if (timeoutId) {
        clearTimeout(timeoutId);
      }
    };
  }, [visible, animate]);

  if (!visible && !animate) return null;
  return (
    <PublishScreenTemplateBlock visible={visible}>
      <Wrapper>
        <Pane>{left}</Pane>
        <Separator />
        <RightPane>{right}</RightPane>
      </Wrapper>
      <HideScroll />
    </PublishScreenTemplateBlock>
  );
};

export default PublishScreenTemplate;
