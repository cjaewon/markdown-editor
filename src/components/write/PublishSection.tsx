import * as React from 'react';
import styled from 'styled-components';
import palette from '../../lib/styles/palette';

const PublishSectionBlock = styled.section`
  & > h3 {
    font-size: 1.3125rem;
    color: ${palette.gray8};
    line-height: 1.5;
    margin-bottom: 0.5rem;
    margin-top: 0;
  }
  .contents {
  }

  & + & {
    margin-top: 1.5rem;
  }
`;

export interface PublishSectionProps extends React.HTMLProps<HTMLElement> {
  title: string;
  children: React.ReactNode;
}

const PublishSection: React.FC<PublishSectionProps> = ({
  title,
  children,
  ...rest
}) => {
  const htmlProps = rest as any;
  return (
    <PublishSectionBlock {...htmlProps}>
      <h3>{title}</h3>
      <div className="contents">{children}</div>
    </PublishSectionBlock>
  );
};

export default PublishSection;
