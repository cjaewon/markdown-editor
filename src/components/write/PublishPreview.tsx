import React from 'react';
import styled, { css } from 'styled-components';
import PublishSection from './PublishSection';
import palette from '../../lib/styles/palette';
import { ImageVector } from '../../static/svg';
import { ellipsis } from '../../lib/styles/utils';

const PublishPreviewBlock = styled(PublishSection)``;

const ThumbnailSizer = styled.div`
  width: 100%;
  padding-top: 55.11%;
  position: relative;
`;

const ThumbnailBlock = styled.div`
  width: 100%;
  height: 100%;
  position: absolute;
  left: 0;
  top: 0;
  box-shadow: 0 0 4px 0 rgba(0, 0, 0, 0.03);
`;

const MissingThumbnail = styled.div`
  background: ${palette.gray2};
  width: 100%;
  height: 100%;
  display: flex;
  justify-content: center;
  align-items: center;
  flex-direction: column;
`;

const Image = styled.img`
  width: 100%;
  height: 100%;
  display: block;
  object-fit: cover;
`;

const UploadButton = styled.button`
  margin-top: 1rem;
  padding: 0.25rem 2rem;
  background: white;
  border-radius: 4px;
  box-shadow: 0px 0px 4px rgba(0, 0, 0, 0.025);
  font-size: 1rem;
  line-height: 1.5;
  color: ${palette.teal5};
  outline: none;
  border: none;
  cursor: pointer;
  transition: all 0.125s ease-in;
  font-weight: bold;
  &:focus {
    box-shadow: 0px 0px 4px rgba(0, 0, 0, 0.15);
  }
  &:hover {
    background: ${palette.gray0};
  }
`;

const PostInfo = styled.div`
  margin-top: 1.5rem;
  h4 {
    line-height: 1.5;
    margin: 0;
    display: block;
    font-size: 1.125rem;
    ${ellipsis};
  }
`;

const ShortDescriptionTextarea = styled.textarea`
  resize: none;
  width: 100%;
  border: none;
  outline: none;
  box-shadow: 0 0 4px 0 rgba(0, 0, 0, 0.03);
  line-height: 1.5;
  font-size: 0.875rem;
  height: 7.375rem;
  padding: 0.75rem 1rem;
  margin-top: 0.5rem;
`;

const TextLimit = styled.div<{ limit: boolean }>`
  text-align: right;
  margin-top: 0.25rem;
  font-size: 0.75rem;
  color: ${palette.gray6};
  ${props =>
    props.limit &&
    css`
      color: ${palette.red6};
    `}
`;

interface ThumbnailProps {
  thumbnail: string | null;
  onUploadClick: () => void;
}
const Thumbnail: React.FC<ThumbnailProps> = ({ onUploadClick, thumbnail }) => {
  return (
    <ThumbnailSizer>
      <ThumbnailBlock>
        {thumbnail ? (
          <Image src={thumbnail} data-testid="image" />
        ) : (
          <MissingThumbnail>
            <ImageVector />
            <UploadButton onClick={onUploadClick}>썸네일 업로드</UploadButton>
          </MissingThumbnail>
        )}
      </ThumbnailBlock>
    </ThumbnailSizer>
  );
};

const ThumbnailModifyBlock = styled.div`
  display: flex;
  justify-content: flex-end;
  margin-bottom: 0.5rem;
  & > .actions {
    display: flex;
    align-items: center;
    button {
      background: none;
      outline: none;
      border: none;
      font-size: 1rem;
      color: ${palette.gray6};
      cursor: pointer;
      padding: 0;
      text-decoration: underline;
      &:hover {
        color: ${palette.gray5};
      }
      &:active {
        color: ${palette.gray7};
      }
    }
    .middledot {
      margin-left: 0.5rem;
      margin-right: 0.5rem;
      display: block;
      width: 2px;
      height: 2px;
      border-radius: 1px;
      background: ${palette.gray6};
    }
  }
`;
type ThumbnailModifyProps = {
  visible: boolean;
  onResetThumbnail: () => void;
  onChangeThumbnail: () => void;
};
const ThumbnailModify: React.FC<ThumbnailModifyProps> = ({
  visible,
  onResetThumbnail,
  onChangeThumbnail,
}) => {
  if (!visible) return null;
  return (
    <ThumbnailModifyBlock>
      <div className="actions">
        <button onClick={onChangeThumbnail}>재업로드</button>
        <div className="middledot" />
        <button onClick={onResetThumbnail}>제거</button>
      </div>
    </ThumbnailModifyBlock>
  );
};

const { useCallback } = React;

export interface PublishPreviewProps {
  title: string;
  description: string;
  thumbnail: string | null;
  onChangeDescription: (description: string) => void;
  onResetThumbnail: () => void;
  onUpload: () => any;
}

const PublishPreview: React.FC<PublishPreviewProps> = ({
  title,
  thumbnail,
  description,
  onChangeDescription,
  onUpload,
  onResetThumbnail,
}) => {
  // const descriptionToShow: string = description || defaultDescription;
  const onChange = (e: React.ChangeEvent<HTMLTextAreaElement>) => {
    onChangeDescription(e.target.value);
  };

  const onKeyDown = useCallback(
    (e: React.KeyboardEvent<HTMLTextAreaElement>) => {
      if (e.keyCode === 13) {
        e.preventDefault();
      }
    },
    [],
  );
  return (
    <PublishPreviewBlock title="포스트 미리보기">
      <ThumbnailModify
        visible={!!thumbnail}
        onResetThumbnail={onResetThumbnail}
        onChangeThumbnail={onUpload}
      />
      <Thumbnail onUploadClick={onUpload} thumbnail={thumbnail} />
      <PostInfo>
        <h4>{title}</h4>
        <ShortDescriptionTextarea
          placeholder="당신의 포스트를 짧게 소개해보세요."
          value={description}
          onChange={onChange}
          onKeyDown={onKeyDown}
        />
        <TextLimit limit={description.length === 150}>
          {description.length}/150
        </TextLimit>
      </PostInfo>
    </PublishPreviewBlock>
  );
};

export default PublishPreview;
