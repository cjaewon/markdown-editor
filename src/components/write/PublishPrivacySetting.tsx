import * as React from 'react';
import styled, { css } from 'styled-components';
import PublishSection from './PublishSection';
import palette from '../../lib/styles/palette';
import { GlobeIcon, LockIcon } from '../../static/svg';

const PublishPrivacySettingBlock = styled(PublishSection)`
  outline: none;
  border: none;
  .contents {
    display: flex;
  }
`;

const Button = styled.button<{ active: boolean }>`
  outline: none;
  border: none;
  flex: 1;
  height: 3rem;
  display: inline-flex;
  align-items: center;
  justify-content: flex-start;
  font-weight: bold;
  background: white;
  font-size: 1.125rem;
  color: ${palette.gray6};
  box-shadow: 0 0 4px 0 rgba(0, 0, 0, 0.05);
  padding: 0;
  padding-left: 1rem;
  border-radius: 4px;
  cursor: pointer;
  border: solid 1px transparent;
  &:hover {
    background: #fdfdfd;
  }
  ${props =>
    props.active &&
    css`
      border: solid 1px ${palette.teal5};
      color: ${palette.teal5};
    `}

  svg {
  }
  & + & {
    margin-left: 1rem;
  }
  .description {
    flex: 1;
    display: flex;
    justify-content: center;
    align-items: center;
  }
`;

export interface PublishPrivacySettingProps {
  isPrivate: boolean;
  onSelect: (value: boolean) => any;
}

const { useCallback } = React;

const PublishPrivacySetting: React.FC<PublishPrivacySettingProps> = ({
  isPrivate,
  onSelect,
}) => {
  const onClickPrivate = useCallback(() => {
    onSelect(true);
  }, [onSelect]);
  const onClickPublic = useCallback(() => {
    onSelect(false);
  }, [onSelect]);
  return (
    <PublishPrivacySettingBlock title="공개 설정">
      <Button active={!isPrivate} onClick={onClickPublic}>
        <GlobeIcon />
        <div className="description">전체 공개</div>
      </Button>
      <Button active={isPrivate} onClick={onClickPrivate}>
        <LockIcon />
        <div className="description">비공개</div>
      </Button>
    </PublishPrivacySettingBlock>
  );
};

export default PublishPrivacySetting;
