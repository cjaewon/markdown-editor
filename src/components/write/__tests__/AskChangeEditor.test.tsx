import * as React from 'react';
import { render, fireEvent } from '@testing-library/react';
import AskChangeEditor, { AskChangeEditorProps } from '../AskChangeEditor';
import { WriteMode } from '../../../modules/write';

describe('AskChangeEditor', () => {
  const setup = (props: Partial<AskChangeEditorProps> = {}) => {
    const initialProps: AskChangeEditorProps = {
      visible: true,
      onCancel: () => undefined,
      onConfirm: () => undefined,
      convertTo: WriteMode.MARKDOWN,
    };
    const utils = render(<AskChangeEditor {...initialProps} {...props} />);
    return {
      ...utils,
    };
  };
  it('renders properly', () => {
    setup();
  });
  it('matches snapshot', () => {
    const { container } = setup();
    expect(container).toMatchSnapshot();
  });
  describe('buttons', () => {
    it('onCancel is working', () => {
      const onCancel = jest.fn();
      const { getByText } = setup({
        onCancel,
      });
      const cancelButton = getByText('취소');
      fireEvent.click(cancelButton);
      expect(onCancel).toBeCalled();
    });

    it('onConfirm is working', () => {
      const onConfirm = jest.fn();
      const { getByText } = setup({
        onConfirm,
      });
      const cancelButton = getByText('확인');
      fireEvent.click(cancelButton);
      expect(onConfirm).toBeCalled();
    });
  });

  describe('convertTo', () => {
    it('convert to markdown', () => {
      const { getByText } = setup({
        convertTo: WriteMode.MARKDOWN,
      });
      getByText('마크다운 에디터로 전환');
    });
    it('convert to quill', () => {
      const { getByText } = setup({
        convertTo: WriteMode.WYSIWYG,
      });
      getByText('쉬운 에디터로 전환');
    });
  });
});
