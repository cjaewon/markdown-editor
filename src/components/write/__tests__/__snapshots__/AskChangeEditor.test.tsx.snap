// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AskChangeEditor matches snapshot 1`] = `
<div>
  <div
    class="sc-bdVaJa kNCPvv"
  />
  <div
    class="sc-bwzfXH fIoYbF"
  >
    <div
      class="sc-htpNat gFFQdP"
    >
      <div
        class="sc-ifAKCX lbyxTT"
      >
        <h3>
          마크다운 에디터로 전환
        </h3>
        <div
          class="message"
        >
          에디터 모드를 전환하시겠습니까?
        </div>
        <div
          class="button-area"
        >
          <button
            class="sc-bxivhb lhCsgH"
            color="lightGray"
          >
            취소
          </button>
          <button
            class="sc-bxivhb jvTLUf"
            color="teal"
          >
            확인
          </button>
        </div>
      </div>
    </div>
  </div>
</div>
`;
